### PR TITLE
feat: support in-memory implementations for keystore provider

### LINKED ISSUE #12: Replace Keystore Provider mock with in-memory implementation in kms-rest

### LINKED ISSUE #15: Refactor to support KMS unlocking on per-keystore basis

### PR DESCRIPTION
**Pull request by aholovko**
Adds support for KMS unlocking on per-keystore basis.
Switches to in-memory implementations for keystore provider in kms-rest.

Closes #12
Closes #15

Signed-off-by: Andriy Holovko <andriy.holovko@gmail.com>

### CODE REVIEW

File: cmd/kms-rest/startcmd/start.go [hunk 1]
```diff
@@ -188,3 +209,100 @@ func healthCheckHandler(rw http.ResponseWriter, r *http.Request) {
...
+func prepareKMSCreator(_ *kmsRestParameters) provider.KMSCreator {
+	return func(keystoreID string) (kms.KeyManager, error) {
+		kmsStorageProv := ariesmemstorage.NewProvider()
+
+		masterKeyReader, err := prepareMasterKeyReader(kmsStorageProv)
+		if err != nil {
+			return nil, err
+		}
+
+		secretLockService, err := local.NewService(masterKeyReader, nil)
```
Comment (Baha-sk) [inline cmd/kms-rest/startcmd/start.go:266]: `nil` argument means the master key is not encrypted, ie: master key lock is nil.

Master keys to be encrypted as well. 

See example of master key lock in the aries-framework-go unit tests [here](https://github.com/hyperledger/aries-framework-go/blob/master/pkg/secretlock/local/local_secret_lock_test.go#L203)

File: cmd/kms-rest/startcmd/start.go [hunk 2]
```diff
@@ -188,3 +209,100 @@ func healthCheckHandler(rw http.ResponseWriter, r *http.Request) {
...
+	masterKeyStore, err := kmsStorageProvider.OpenStore(masterKeyStoreName)
+	if err != nil {
+		return nil, err
+	}
+
+	masterKey, err := masterKeyStore.Get(masterKeyDBKeyName)
+	if err != nil {
+		if errors.Is(err, ariesstorage.ErrDataNotFound) {
+			const size = 32
+			masterKeyRaw := random.GetRandomBytes(uint32(size))
```
Comment (Baha-sk) [inline cmd/kms-rest/startcmd/start.go:295]: use crypto.Rand here

File: cmd/kms-rest/startcmd/start.go [hunk 3]
```diff
@@ -188,3 +209,100 @@ func healthCheckHandler(rw http.ResponseWriter, r *http.Request) {
...
+	}
+
+	return keystoreProvider{
+		storageProvider: memstore.NewProvider(),
+		kmsCreator:      prepareKMSCreator(parameters),
+		crypto:          c,
+	}, nil
+}
+
+func prepareKMSCreator(_ *kmsRestParameters) provider.KMSCreator {
```
Comment (DRK3) [inline cmd/kms-rest/startcmd/start.go:257]: Is there a reason for the unused `kmsRestParameters`?

File: pkg/keystore/keystore.go [hunk 4]
```diff
@@ -50,15 +53,20 @@ type Keystore struct {
 }
 
 // New returns a new instance of keystore.
-func New(keystoreID string, provider Provider) (*Keystore, error) {
+func New(keystoreID string, provider provider.Provider) (*Keystore, error) {
```
Comment (DRK3) [inline pkg/keystore/keystore.go:56]: Consider avoiding the repetition of names - "provider.Provider". Perhaps a different package name?

(sometimes the linter will complain about this)

File: pkg/provider/provider.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+package provider
+
+import (
+	"github.com/hyperledger/aries-framework-go/pkg/crypto"
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents provider with functionality needed for keystore.
+type Provider interface {
```
Comment (DRK3) [inline pkg/provider/provider.go:16]: I wonder if there's another name we can come up with for this interface to make it more distinguishable from other "Provider" types

File: cmd/kms-rest/startcmd/start.go (hunk 1 above)
Comment (aholovko) [reply cmd/kms-rest/startcmd/start.go:192]: I started adding support for masterkey lock but it requires quite a lot of changes (especially logic related to getting user-specific passphrase) so decided to delegate that to the new ticket - https://github.com/trustbloc/hub-kms/issues/17

File: cmd/kms-rest/startcmd/start.go (hunk 3 above)
Comment (aholovko) [reply cmd/kms-rest/startcmd/start.go]: indeed, we don't need it now, so removed

File: pkg/keystore/keystore.go (hunk 4 above)
Comment (aholovko) [reply pkg/keystore/keystore.go:56]: looks like it's quite common to name "main" type in the package after the name of the package itself (e.g. `crypto.Crypto`, `did.DID` from Aries framework, `template.Template`, `rand.Rand`, `image.Image` from Go standard library) but I do agree that here we may try to come up with a better name. I am planning some additional refactoring (as part of https://github.com/trustbloc/hub-kms/issues/17), so maybe something better will emerge as part of it (but can't promise).

File: pkg/provider/provider.go (hunk 5 above)
Comment (aholovko) [reply pkg/provider/provider.go]: addressed it in the comment above (will consider to come up with a better name in upcoming refactoring)

File: cmd/kms-rest/startcmd/start.go (hunk 2 above)
Comment (aholovko) [reply cmd/kms-rest/startcmd/start.go]: did you mean using `rand.Read` from the standard `crypto/rand` package as in snippets below?

```go
func generateRandomBytes() []byte {
	keySize := sha256.New().Size()
	buf := make([]byte, keySize)

	_, err := rand.Read(buf)
	if err != nil {
		panic(err) // out of randomness, should never happen :-)
	}

	return buf
}
```

Comment (Baha-sk) [reply cmd/kms-rest/startcmd/start.go]: yes

File: cmd/kms-rest/startcmd/start.go [hunk 6]
```diff
@@ -94,78 +121,117 @@ func createStartCmd(srv server) *cobra.Command {
...
+	if err != nil {
+		return nil, err
+	}
 
-	return srv.ListenAndServe(parameters.hostURL, constructCORSHandler(router))
+	masterKey, err := masterKeyStore.Get(masterKeyDBKeyName)
+	if err != nil {
+		if errors.Is(err, ariesstorage.ErrDataNotFound) {
+			masterKeyContent := random.GetRandomBytes(keySize)
+			masterKey = []byte(base64.URLEncoding.EncodeToString(masterKeyContent))
```
Comment (Baha-sk) [inline cmd/kms-rest/startcmd/start.go:223]: fyi, this line sets an `unencrypted` master key to be stored in the line below.

when master lock is created in #17, you need to update this line to set the `encrypted` master key in the store

Comment (aholovko) [reply cmd/kms-rest/startcmd/start.go:223]: yep, I noticed that already - without updating that line the following error happens: "cipher: message authentication failed"

thank you @Baha-sk for clarifying that part

File: pkg/keystore/keystore.go (hunk 4 above)
Comment (DRK3) [reply pkg/keystore/keystore.go:56]: OK, that sounds reasonable to me. I think if an idea comes to mind then perhaps you could consider it, but it looks like you're right about this being common, so I think it's not much of a concern for me anymore

File: pkg/provider/provider.go (hunk 5 above)
Comment (DRK3) [reply pkg/provider/provider.go]: As an example: in EDV, I have an "edvprovider" that extends a provider to provide additional functionality. Perhaps a similar naming scheme can work here

File: cmd/kms-rest/startcmd/start.go (hunk 2 above)
Comment (aholovko) [reply cmd/kms-rest/startcmd/start.go]: done cc @Baha-sk

**Pull request merged**